sklearn/preprocessing/data.py

```diff
@@ -1956,3 +1967,222 @@ def transform(self, X):
+
+
+class CountFeaturizer(BaseEstimator, TransformerMixin):
+
+    """
+    Adds in a new feature column containing the number of occurences of
+    each row in the given data set.
+
+    Formally, for each data point X_i in the dataset X, it will add in
+    a new feature count_X_i to the end of X_i where count_X_i is the number of
+    occurences of X_i in the dataset X given the equality indicator
+    'inclusion'
+
+    This preprocessing step is useful when the number of occurences
+    of a particular piece of data is helpful in computing the prediction.
+
+    Parameters
+    ----------
+    inclusion: set, list, numpy.ndarray, or string (only 'all'), default='all'
+        The inclusion criteria for counting
+        - 'all' (default): Every feature is taken into consideration when
+        counting how many of each row is in the dataset.
+        - collection of indices: A collection of values where if the index v is
+        in 'inclusion', then the vth feature will be put into the equality
+        checks when counting how often each row occurs.
+
+    removal_policy: set, list, numpy.ndarray, or None, default=None
+        - None (default): No features are removed on transformation.
+        - 'inclusion': Every feature counted will be removed and replaced with
+        the count feature. This is equivalent to having the 'inclusion' and
+        'removal_policy' parameter being set to the exact same collection.
+        - collection of indices: For each index v in 'removal_policy', the vth
+        feature will be removed in the transformation (and be replaced with
+        the new count feature).
+
+    Examples
+    --------
+    Given a dataset with two features and four samples, we let the transformer
+    find the number of occurences of each data point in the dataset
+
+    >>> from sklearn.preprocessing.data import CountFeaturizer
+    >>> data = [[1, 1], [1, 1], [3, 1], [0, 0]]
+    >>> cf = CountFeaturizer()
+    >>> cf.fit_transform(data)      # doctest: +NORMALIZE_WHITESPACE
+    array([[ 1.,  1.,  2.],
+           [ 1.,  1.,  2.],
+           [ 3.,  1.,  1.],
+           [ 0.,  0.,  1.]])
+    >>> cf = CountFeaturizer(inclusion=[1])
+    >>> cf.fit_transform(data)      # doctest: +NORMALIZE_WHITESPACE
+    array([[ 1.,  1.,  3.],
+           [ 1.,  1.,  3.],
+           [ 3.,  1.,  3.],
+           [ 0.,  0.,  1.]])
+    >>> data_2 = [[1, 1]]
+    >>> cf = CountFeaturizer()
+    >>> cf.fit(data).transform(data_2)
+    array([[ 1.,  1.,  2.]])
+
+    Notice how on the second fit, we set the inclusion to [0, 1]
+    which made it only count the number of instances of the second feature
+
+    See also
+    --------
+    WIP, I will make the documentation .rst afterwards
+    """
+
+    def __init__(self, inclusion='all', removal_policy=None):
+        # removal policy currently not implemented, but the method
+        # signatures are set
+        CountFeaturizer._check_params(inclusion=inclusion,
```

Comment:
Our convention is to perform validation and cleaning of parameters in `fit`. `__init__` should consist only of setting member variables. Please read the contributors' guide.